### PR TITLE
Hog/patch one

### PR DESCRIPTION
**Pull request by ericlee0803**
A set of small changes to enable early stopping in the XGB integration

Changes:

- Change the XGB experiment API to support an additional argument, early_stopping_rounds, which is simply passed into the runs API (which already supports this argument). 
- Unlike the runs API, early_stopping_rounds is active and set to ten by default. If it is None, then it is deactivated. 
- If early_stopping_rounds is active, then we record an additional metric: "num_boost_round_before_stopping", which is the number of boosting rounds actually performed before early stopping kicks in. 
- Also added an integration test to confirm its functionality. 

Still need to do:

- Update documentation to reflect these changes (will be done after this PR is merged in)

### CODE REVIEW

**Comment by ericlee0803** [comment]
Whoops, let me make additional changes to the PR and then I will issue this again

File: sigopt/xgboost/experiment.py [hunk 1]
```diff
@@ -181,14 +183,31 @@ def run_experiment(self):
...
 
-def experiment(experiment_config, dtrain, evals, params, num_boost_round=None, run_options=None):
+def experiment(
+  experiment_config,
+  dtrain,
+  evals,
+  params,
+  num_boost_round=None,
+  run_options=None,
+  early_stopping_rounds=10,
```
Comment (harvey-intel) [inline sigopt/xgboost/experiment.py:199]: flip the order of `early_stopping_rounds` and `run_options`?

File: sigopt/xgboost/run.py [hunk 2]
```diff
@@ -278,10 +278,15 @@ def train_xgb(self):
...
     # Always log xgb-default eval_metric
+    n_eval_rounds = 0
     if self.evals_result is not None:
       for dataset, metric_dict in self.evals_result.items():
         for metric_label, metric_record in metric_dict.items():
           self.run.log_metric(f"{dataset}-{metric_label}", metric_record[-1])
+          n_eval_rounds = len(metric_record)
+
+      if self.early_stopping_rounds:
+        self.run.log_metric('num_boost_round_before_stopping', n_eval_rounds)
```
Comment (harvey-intel) [inline sigopt/xgboost/run.py:289]: You can get this much easier with just the `booster.best_iteration` call. Which should probably be what we call it as well.

**Review by harvey-intel — CHANGES_REQUESTED** [review]
I think there is a glaring problem with this PR. The metric we are using for the early stopping criteria on can be different from the metric we are optimizing, which seems confusing here.

For example, the default `eval_metric` for classification is `logloss` in xgboost. But with our integration, the default optimized metric for classification is `accuracy`.  We need to somehow reconcile these two.

**Comment by ericlee0803** [comment]
You make a good point. I never gave much thought to the discrepancy between the metric optimized by BO, and the loss function used to train the XGB model. 

We allow this discrepancy right now during an experiment api, which defaults to optimize accuracy in the outer BO loop but default to logloss as the XGB training metric. So I do think it's fine to extend this discrepancy to early stopping as well... early stopping on the training metric seems reasonable to me, just a bit ambiguous.  

I agree that logically consistent thing to do is early stop on the BO outer loop metric. I can patch that in, it will just take additional time and I was hoping to get this in by the end of the month so I can run some experiments for Mike and Marketing to use. 

Perhaps we can early stop on the training metric for now, then I can issue a later patch once I put in the optimize eval_metrics in the outer BO loop patch. Once I do that, it'll be much easier to do what you're suggesting. Thoughts?

File: sigopt/xgboost/run.py (hunk 2 above)
Comment (ericlee0803) [reply sigopt/xgboost/run.py:289]: I think the best_iteration is not what we want, if the loss / eval_metric drops later, then best_iteration won't match num_boost_round_before_stopping

File: sigopt/xgboost/experiment.py (hunk 1 above)
Comment (ericlee0803) [reply sigopt/xgboost/experiment.py:199]: done

File: integration_test/test_xgboost_experiment.py [hunk 3]
```diff
@@ -104,6 +104,20 @@ def test_experiment(self, task):
...
+    self.experiment_params['early_stopping_rounds'] = 1
+    experiment = sigopt.xgboost.experiment(**self.experiment_params)
+    assert experiment.is_finished()
+    sigopt_suggested_runs = list(experiment.get_runs())
+    for ssr in sigopt_suggested_runs:
+      assert ssr.assignments['early_stopping_rounds']
+      for param in experiment.parameters:
+        if param.name == 'num_boost_round':
+          assert ssr.values['num_boost_round_before_stopping']
+          assert ssr.assignments['num_boost_round'] >= ssr.assignments['num_boost_round_before_stopping']
```
Comment (yuhong-sigopt) [inline integration_test/test_xgboost_experiment.py:119]: Based on the assertion, we can not make sure whether it is actually early stopped or not.

File: sigopt/xgboost/run.py (hunk 2 above)
Comment (harvey-intel) [reply sigopt/xgboost/run.py:289]: Am I misunderstanding this from the xgboost doc? 
https://xgboost.readthedocs.io/en/stable/python/python_intro.html#early-stopping
```
 If early stopping occurs, the model will have three additional fields: bst.best_score, bst.best_iteration.
```

**Comment by harvey-intel** [comment]
> We allow this discrepancy right now during an experiment api, which defaults to optimize accuracy in the outer BO loop but default to logloss as the XGB training metric. So I do think it's fine to extend this discrepancy to early stopping as well... early stopping on the training metric seems reasonable to me, just a bit ambiguous.

The XGB models are trained on the objective (loss function), not the evaluation metrics. So the discrepancy before is fine. 

> I agree that logically consistent thing to do is early stop on the BO outer loop metric.

I think there are two ways to approach this.

1. When early stopping is active, change the optimized metric (in the experiment config) to the evaluation metric that's being used for early stopping.
2. Use a custom early stopping callback function and early stop on the optimized metric. This is slightly more complicated, but I think makes more sense from the user perspective.

File: integration_test/test_xgboost_experiment.py (hunk 3 above)
Comment (harvey-intel) [reply integration_test/test_xgboost_experiment.py:119]: Yea, I don't think there is any way to guarantee the model early stopped.

File: sigopt/xgboost/run.py (hunk 2 above)
Comment (ericlee0803) [reply sigopt/xgboost/run.py:289]: The docs say: "_Note that xgboost will return a model from the last iteration, not the best one_", so I think it's implied that best_iteration is not necessarily the final one (which is what num_boost_round_before_stopping records)

Comment (harvey-intel) [reply sigopt/xgboost/run.py:289]: Ok I see. It's probably something like 
```
num_boost_round_before_stopping - best_iteartion <= early_stopping_round
```
We should record both `best_iteration` and `num_boost_round_before_stopping`.  I did find this, which I think is equivalent to your `nbs before stopping`.
https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.Booster.num_boosted_rounds
